src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1365 @@
+                );
+                if self.fixed_stack_slots.contains(preg)
+                    && self.edits.scratch_regs[preg.class()].is_none()
+                {
+                    let reg = self.get_scratch_reg(preg.class())?;
+                    self.edits.scratch_regs[preg.class()] = Some(reg);
+                    self.available_pregs[OperandPos::Early].remove(reg);
+                    self.available_pregs[OperandPos::Late].remove(reg);
+                }
+                self.evict_vreg_in_preg(inst, preg);
+                self.vreg_in_preg[preg.index()] = VReg::invalid();
+            }
+        }
+        self.remove_clobbers_from_available_pregs(clobbers);
+        for preg in clobbers {
+            if self.vreg_in_preg[preg.index()] != VReg::invalid() {
+                trace!(
+                    "Evicting {} from clobber {preg}",
+                    self.vreg_in_preg[preg.index()]
+                );
+                if self.fixed_stack_slots.contains(preg)
+                    && self.edits.scratch_regs[preg.class()].is_none()
+                {
+                    let reg = self.get_scratch_reg(preg.class())?;
+                    self.edits.scratch_regs[preg.class()] = Some(reg);
+                    self.available_pregs[OperandPos::Early].remove(reg);
+                    self.available_pregs[OperandPos::Late].remove(reg);
+                }
+                self.evict_vreg_in_preg(inst, preg);
+                self.vreg_in_preg[preg.index()] = VReg::invalid();
+            }
+        }
+        for (_, op) in operands.non_fixed_use() {
+            if op.as_fixed_nonallocatable().is_some() {
+                continue;
+            }
+            if let Some(preg) = self.vreg_allocs[op.vreg().vreg()].as_reg() {
+                trace!("Removing {op}'s current reg allocation {preg} from reg sets");
+                // The current allocation, vreg_allocs[op.vreg], doesn't change,
+                // so it should be removed from the available reg sets to avoid
+                // allocating it to some other operand in the instruction.
+                //
+                // For example:
+                // 1. def v0 (reuse: 1), use v1, use v2
+                // 2. use v1 (fixed: p0)
+                //
+                // When inst 1 is about to be processed, vreg_allocs[v1] will be p0.
+                // Suppose p1 is allocated to v0: this will create a fixed constraint for
+                // v1 and p1 will also be allocated to it.
+                // When it's time to process the v2 operand, vreg_allocs[v1] will still be p0
+                // because it doesn't change (except by an explicit fixed reg constraint which
+                // will not be a problem here) and it's possible for v2 to get p0 as an allocation,
+                // which is wrong. That will lead to the following scenario:
+                //
+                // move from p0 to p1   // Inserted due to reuse constraints
```

Comment:
I just realized that this loop actually has to be present, even before any evictions occur. An example:

```
1. def v0 (fixed: p23), use v1 (reg)
2. use v1 (fixed: p0)
// v0 is in stack_v0 and v1 is in stack_v1
```

If p23 is a stack slot, then we'll need to insert an edit to move from p23 to stack_v0 after inst 1 and a scratch register has to be used. Since p0 is still marked as available in `available_pregs`, it is possible for it to be used as scratch. If that happens then we'll have:

```
1. def v0 (fixed: p23), use v1 (reg)
move from stack_v1 to p0     // v1 is evicted
move from p23 to p0             // v1 is overwritten by v0
move from p0 to stack_v0
2. use v1 (fixed: p0)                // v0 is used instead of v1
```

Which is incorrect.

Reply:
I think this still works because moves are emitted in reverse. Since the eviction of v1 happens first when a scratch register is selected, the actual move of stack_v1 to p0 will happen last.

Reply:
Right.